proposals/0022-validating-resource-container-elements.md

```diff
@@ -2,30 +2,41 @@
 * Author(s): [Joshua Batista](https://github.com/bob80905)
 * Sponsor: TBD
 * Status: **Under Consideration**
-* Planned Version: Shader Model 6.8
-* PRs:
-* Issues:
-  [#75676](https://github.com/llvm/llvm-project/issues/75676)
+* Impacted Project(s): (LLVM)
+
+*During the review process, add the following fields as needed:*
+
+* PRs: [#NNNN](https://github.com/microsoft/DirectXShaderCompiler/pull/NNNN)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
 
 ## Introduction
 
-There's a specific closed set of types that are valid as element types for an HLSL resource, 
-which includes int/uint of sizes 16/32/64, half, float, and double. If someone writes 
-`RWBuffer<MyCustomType>` we should reject it and give the user a nice diagnostic. 
+There's a specific closed set of types that are valid as element types for an HLSL
+resource, which includes int/uint of sizes 16/32/64, half, float, and double. Arrays of
+these types are not valid. If someone writes `RWBuffer<MyCustomType>` we should reject it
+and give the user a nice diagnostic. 

```

Comment:
There's a key bit of information omitted from here which is that it sounds like you want to support user defined types that are equivalent to one of the supported types.

In the examples you have:

```hlsl
struct x {
	int i;
};

struct bad {
	x a;
	bool b;	
};

RWBuffer<x> valid;
RWBuffer<bad> invalid;
```

But...is it because of the `bool` that is is invalid?  How about these examples:

```hlsl
struct a {
   int aa;
   int ab;
};

struct b {
   x bx;  // (same x as in example above)
};

struct c {
  a ca;
  float4 cb;
};
```

Reply:
Sure, I can add some extra clarification, but if I understand correctly, I believe x should be allowed, bad should not be allowed because of the bool, a should be allowed, as well as b and c.

Reply:
The main thing I'm looking for here is an explanation in the introduction or motivation section on what the rule is actually meant to be.  From the description of that rule I should be able to determine which of those examples are correct.  The examples can be included to help people understand what the rule is, but I don't see anything in the current text saying what the rule is, beyond:

> There's a specific closed set of types that are valid as element types for an HLSL resource, which includes int/uint of sizes 16/32/64, half, float, and double. Arrays of these types are not valid.

From this description none of the struct examples are valid.


Then you can discuss the proposed solution for validating this rule.